### PR TITLE
Optimization change model

### PR DESCRIPTION
**Pull request by almostcake**
Replaced recursive maplayers-logic with detail-array of maplayer
Now MapE projection used for loading map-maplayer-layerlink
Maplayers tree operates with model.hierarchy property which contains sorted by index hierarchy of map,mapLayer
Added extended getter and setter to search objects by path if it contains Ember.RecordArray or other Enumerable
tests are OK, functionality too

### CODE REVIEW

File: tests/dummy/app/templates/map.hbs [hunk 1]
```diff
@@ -42,44 +42,44 @@
...
-    </div>
+          add=(action "onMapLayerAdd" "model.hierarchy")
+        }}
+      </div>
     <div class="row">
       <div class="sixteen wide column">
         {{#flexberry-maptoolbar
           leafletMap=leafletMap
-          layers=model.rootLayer.layers as |mapToolbar|
+          layers=model.hierarchy as |mapToolbar|
```
Comment (kuzkok) [inline tests/dummy/app/templates/map.hbs]: Why not mapLayer?

Comment (almostcake) [inline tests/dummy/app/templates/map.hbs]: thought identity and search tools are already set up to work with hierarchical layers, so I decided not to change, cuz model.hierarchy equals model.rootLayer.layers

Comment (kuzkok) [inline tests/dummy/app/templates/map.hbs]: ok

File: addon/components/base-layer.js [hunk 2]
```diff
@@ -170,7 +170,7 @@ export default Ember.Component.extend(
     _query(e) {
       let layerLinks = this.get('layer.layerLink');
```
Comment (kuzkok) [inline addon/components/base-layer.js]: layer was renamed to layerModel

Comment (almostcake) [inline addon/components/base-layer.js]: fixed, it was not obvious while merging :)

**Pull request merged**